### ISSUE TITLE
Dimension mismatch warning for loss calculation.

### ISSUE THREAD

**Issue by zsameem** [sefibk/KernelGAN#26]
Hi,

Pytorch gives the following warnings when I run train.py for kernel estimation.

```
(venv) samim@joe:~/Desktop/ms/KernelGAN$ python train.py --input-dir ./images/ --SR --real
Scale Factor: X2 	ZSSR: True 	Real Image: True
************************************************************
STARTED KernelGAN on: "./images/KLE_9806.png"...
  0%|                              | 0/3000 [00:00<?, ?it/s]/home/samim/Desktop/ms/resLF/venv/lib/python3.7/site-packages/torch/nn/modules/loss.py:88: UserWarning: Using a target size (torch.Size([13])) that is different to the input size (torch.Size([1, 1, 13, 13])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.
  return F.l1_loss(input, target, reduction=self.reduction)
/home/samim/Desktop/ms/resLF/venv/lib/python3.7/site-packages/torch/nn/modules/loss.py:88: UserWarning: Using a target size (torch.Size([])) that is different to the input size (torch.Size([1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.
  return F.l1_loss(input, target, reduction=self.reduction)
/home/samim/Desktop/ms/resLF/venv/lib/python3.7/site-packages/torch/nn/modules/loss.py:431: UserWarning: Using a target size (torch.Size([2])) that is different to the input size (torch.Size([2, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.
  return F.mse_loss(input, target, reduction=self.reduction)
100%|███████████████████| 3000/3000 [01:27<00:00, 34.35it/s]
KernelGAN estimation complete!
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running ZSSR X2...

ZSSR configuration is for a real image
```

It seems that there is a mismatch in the dimensions. Is it by design or is it not supposed to happen. 

I also have another question about doing SR after kernel estimation using ZSSR. Does the ZSSR get trained on the input images or do you use a pretrained ZSSR somehow?

Best regards

**Comment by sefibk** [comment]
Hi,
I don't recall these warnings, before I go about checking them, please verify packages version (specifically Pytorch) and that your image size is divisible by 2 (better by 4). Please let me know if this was the issue.
Regarding ZSSR - similarly to KernelGAN it is an image specific NN that trains on the input image only. Here KernelGAN uses ZSSR to train on the specific image and KERNEL! => No pre-training for ZSSR

**Comment by zsameem** [comment]
Hi,
The size of the image I used was divisible by 4 (2178 x 1454). I am using torch==1.2.0, torchvision==0.4.0.
Thanks for the info on how ZSSR works.

Also, I used this for blind SR of a real image. There were no crazy artifacts in the output image but at the same time it was also very close to a naive upsampling from gimp. Any tips on getting better results on real images would be helpful?

**Comment by sefibk** [comment]
I used torch 1.0.0, try switching to that and see if you still get the warnings. Any way, if you didn't change configurations then I don't recall this warning.

Regarding the SR quality - some images are much more 'difficult' to do SR. It could be for several reasons e.g. type of camera, motion blur, the single kernel assumption doesn't hold etc. Our method works well in many cases but not all.

**Comment by zsameem** [comment]
Thanks for all the information.
